### PR TITLE
Add ability to use paid maxmind service

### PR DESCRIPTION
**Pull request by remarcable**

### CODE REVIEW

File: lib/getLocationForIpWithService.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { Meteor } from 'meteor/meteor';
+import Geoip2ws from 'geoip2ws';
+
+import { sanitizeServiceResult } from './sanitizeResult';
+
+const userId = Meteor.settings.maxmind.userId;
```
Comment (nerdmed) [inline lib/getLocationForIpWithService.js]: unsafe access, why error if there are no settings or no maxmind namespace?

File: lib/asyncHTTP.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { HTTP } from 'meteor/http';
+
+function asyncHTTP(url, method = 'GET', options = {}) {
```
Comment (nerdmed) [inline lib/asyncHTTP.js]: why is this needed? can we use a fetch npm module? It should return a promise.
This function is only used once. I would suggest to just use a promise down there, or just use the callback to KIS 👍

File: lib/getLocationForIpWithService.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+const licenseKey = Meteor.settings.maxmind.licenseKey;
+
+const lookupIpWithService = new Geoip2ws({
+    userId,
+    licenseKey,
+    service: 'country',
```
Comment (nerdmed) [inline lib/getLocationForIpWithService.js]: can i target the service?

File: Readme.md [hunk 4]
```diff
@@ -25,34 +24,53 @@ See http://dev.maxmind.com/geoip/legacy/csv/ for a more exact rundown on some of
...
+        "userId": "YOUR_USER_ID",
+        "licenseKey": "YOUR_LICENSE_KEY"
+    }
+}
+```
 
-See the `clientAddress` section at http://docs.meteor.com/#/full/meteor_onconnection for more information.
+You can then call a Meteor method to retrieve the data:
+```js
+Meteor.call('UserLocation/getForIpWithService', ip, callback);
```
Comment (nerdmed) [inline Readme.md]: witch service? What do you think about making this clear: `loopUpWithMaxmind`or something similar: `getIpWithMaxmind`

File: lib/getLocationForIpWithService.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    requestTimeout: 2000,
+});
+
+export default function getLocationForIpWithService(ip) {
+    return new Promise((resolve, reject) => {
+        if (!userId || !licenseKey) {
```
Comment (nerdmed) [inline lib/getLocationForIpWithService.js]: confused: thought this is the meteor userId - maybe we can name it maxmindUserId or something

File: lib/getLocationForIpWithService.js (hunk 3 above)
Comment (nerdmed) [reply lib/getLocationForIpWithService.js]: suggestion - make service an option with default: maxmindLookup({ service = 'country' }),

File: lib/getUserIp.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+const GET_IP_API_ENDPOINT = 'http://bot.whatismyipaddress.com/';
+const localhostIpAddress = '127.0.0.1';
+
+export default function getUserIp(meteorConnection) {
+    return new Promise((resolve, reject) => {
+        const clientIp = meteorConnection.clientAddress;
+        const isLocalIp = !clientIp || clientIp === localhostIpAddress;
+
+        if (Meteor.isDevelopment && isLocalIp) {
+            asyncHTTP(GET_IP_API_ENDPOINT)
```
Comment (nerdmed) [inline lib/getUserIp.js]: why is this needed? Only for development?

File: lib/sanitizeResult.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+            correctedResult.ip = ip;
+        } catch (e) {
+            reject('Error performing lookup function', e);
+        }
+
+        resolve(correctedResult);
+    });
+}
+
+export function sanitizeServiceResult(result) {
```
Comment (nerdmed) [inline lib/sanitizeResult.js]: which service? I would suggest to name it to make it clear

File: lib/sanitizeResult.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    return new Promise((resolve, reject) => {
```
Comment (nerdmed) [inline lib/sanitizeResult.js]: why is this wrapped in a promise?

**Review by nerdmed — CHANGES_REQUESTED** [review]

File: lib/getLocationForIpWithMaxMind.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+const maxMindSettings = _.defaults(meteorMaxMindSettings, defaultMaxMindSettings);
+const lookupIpWithMaxMind = new Geoip2ws(maxMindSettings);
+
+
+export default function getLocationForIpWithMaxMind(ip, sanitizeResult = true) {
+    return new Promise((resolve, reject) => {
+        const { userId: maxMindUserId, licenseKey } = maxMindSettings;
+        if (!maxMindUserId || !licenseKey) {
+            reject('You didn\'t provide userId or licenseKey for maxmind');
```
Comment (remarcable) [inline lib/getLocationForIpWithMaxMind.js]: pass in error obj

File: Readme.md [hunk 10]
```diff
@@ -25,34 +24,63 @@ See http://dev.maxmind.com/geoip/legacy/csv/ for a more exact rundown on some of
 
 There are two Meteor Methods provided that can be called from the client or server. Both return the same location info as returned by `UserLocation.get()`, documented above.
 
-```
+```js
 Meteor.call('UserLocation/get', callback) // IP lookup for the connected user
 Meteor.call('UserLocation/getForIp', ip, callback) // IP lookup for any ip
```
Comment (remarcable) [inline Readme.md]: getWithMaxMind

File: lib/getLocationForIpLocal.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
+import { Meteor } from 'meteor/meteor';
+import lookupIp from 'geoip-lite';
```
Comment (nerdmed) [inline lib/getLocationForIpLocal.js]: import geoIpLite.lookup()

File: lib/getLocationForIpWithMaxMind.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+
+const meteorMaxMindSettings = Meteor.settings.maxmind || {};
+
+const defaultMaxMindSettings = {
+    service: 'country',
+    requestTimeout: 2000,
+};
+
+const maxMindSettings = _.defaults(meteorMaxMindSettings, defaultMaxMindSettings);
```
Comment (nerdmed) [inline lib/getLocationForIpWithMaxMind.js]: const maxMindSettings = {
...defaultMaxMindSettings
... meteorMaxMindSettings
}

File: user-location.js [hunk 13]
```diff
@@ -1,16 +1,22 @@
...
-    Meteor.call("UserLocation/get", function(e, res){
-        if (!e && res) {
+export const UserLocation = new ReactiveVar({});
+
+UserLocation.update = function (isOnStartUp) {
+    if (!isOnStartUp) {
+        console.warn('[user-location]: Using the ReactiveVar is deprecated and will be removed in the next major release. Use provided meteor methods instead.');
+    }
+
+    Meteor.call('UserLocation/get', (err, res) => {
```
Comment (nerdmed) [inline user-location.js]: das hier in eigene function

Comment (nerdmed) [reply user-location.js]: hier aufrufen

File: user-location.js [hunk 14]
```diff
@@ -1,16 +1,22 @@
...
         }
-    })
-}
+    });
+};
 
-Meteor.startup(function() {
-    UserLocation.update();
+Meteor.startup(() => {
+    UserLocation.update(true);
```
Comment (nerdmed) [inline user-location.js]: die neue function hier direkt callen

**Review by nerdmed — APPROVED** [review]
lgtm

**Pull request merged**